Adding a 'dev mode' question to the security section of the `risc0` FAQ (#295)
risc0/website: docs/faq.md

### zkVM Application Design
<a class="anchor" id="using-receipts"></a>

> j<details closed>

Reviewer: Stray character?
```suggestion
<details closed>
```

<br/>

## Security

> <a class="anchor" id="image-id-security"></a>

Reviewer: IMO this is better categorized under features/perf/limits than security. It certainly has security _implications_ that we want to be very clear about, but it's not really a security feature.

<details closed>
<summary>
Q:
What is developer mode and how can I use it safely?
</summary>
A: We support a development-only mode for standalone risc0 projects in which code is executed but not proven. This adds efficiency to the development cycle during development stages where proving correct execution is not yet critical.

> 

Reviewer: This is rendering oddly for me. Here as well as after the bold paragraph there is a newline but no other gap between paragraphs, whereas the other paragraph breaks have a newline plus a gap. Not sure what's wrong or how to fix it, but if you can spend at least a little while troubleshooting it would be nice to clean this up.

> A risc0 project in dev mode supports (fake) proof receipt creation and pass-through (fake) 'verification' workflows, so that dev mode may be switched on and off without impacting project workflows. In particular, receipts generated in dev mode still include public outputs written to the journal.

Reviewer 2: ```suggestion
A risc0 project in dev mode supports ([fake](https://docs.rs/risc0-zkvm/0.18.0/risc0_zkvm/enum.InnerReceipt.html#variant.Fake)) receipt creation and pass-through (fake) 'verification' workflows, so that dev mode may be switched on and off without impacting project workflows. In particular, receipts generated in dev mode still include public outputs written to the journal.
```

However, because the proving process is bypassed, receipts generated in dev mode will fail a standard receipt verification check. Only when the verifier is also built in dev mode will it perform pass-through 'verification' of the fake receipt.

> <b>To keep this mode out of production environments, we recommend building production-ready projects with the "disable-dev-mode" feature flag; it is absent by default.</b>

Reviewer 2: ```suggestion
<b>To keep this mode out of production environments, we recommend building production-ready projects with the "disable-dev-mode" [feature flag](https://github.com/risc0/risc0/#feature-flags); it is absent by default.</b>
```

Only projects built without this flag may run dev mode. Enabling dev mode requires also that the environment variable "RISC0_DEV_MODE" be set.

> As additional protection, if the dev mode environment variable is present alongside a project built with the "disable-dev-mode" feature flag, the project will panic. The same is true if the feature flag and environment variable are both absent, as this is also an ambiguous signal of intent.

Reviewer 2: ```suggestion
As additional protection, if the dev mode [environment variable](https://docs.rs/risc0-zkvm/0.18.0/risc0_zkvm/fn.is_dev_mode.html) is present alongside a project built with the "disable-dev-mode" feature flag, the project will panic. The same is true if the feature flag and environment variable are both absent, as this is also an ambiguous signal of intent.
```

Reviewer: Hmm, it feels a bit strange to me to link `environment variable` to a page that doesn't mention `RISC0_DEV_MODE` anywhere. What about extending the link?

Also, this is making me wonder if https://github.com/risc0/risc0/pull/909 should be expanded to include updating the content at the link to say more about what dev mode is and when it is and isn't enabled.
```suggestion
As additional protection, if the dev mode [environment variable is present](https://docs.rs/risc0-zkvm/0.18.0/risc0_zkvm/fn.is_dev_mode.html) alongside a project built with the "disable-dev-mode" feature flag, the project will panic. The same is true if the feature flag and environment variable are both absent, as this is also an ambiguous signal of intent.
```

For further reference, take a look at the table below. To learn more about usage, see the <a href="https://github.com/risc0/risc0/README.md">risc0 project README</a>. For a closer look at implementation, take a look at the <a href="https://github.com/risc0/risc0/risc0/zkvm/src/host/server/prove/dev_mode.rs">dev mode source</a>.

| | Feature flag off | Feature flag on|
| --- | --- | --- |
| RISC0_DEV_MODE=true| Developer mode activated | Prover panic |

> | RISCO_DEV_MODE={false, unset}| Prover panic | Default project behavior |

Reviewer: I think that the "no `RISC0_DEV_MODE`, no `-F disable-dev-mode`" generates proofs, although I may be mistaken
```suggestion
| RISCO_DEV_MODE={false, unset}| Default project behavior | Default project behavior |
```
